Modules.md

```diff
@@ -170,30 +231,154 @@ reserve up to the *maximum* is not. When a maximum memory size is *not* declared
+of table definitions is limited to 1, but this may be relaxed in the
+[future](FutureFeatures.md#multiple-tables-and-memories).
+
+A table definition may declare itself to be the
+[default](AstSemantics.md#table) table of the module. At most
+one table definition may declare itself to be the default. In the MVP,
+if there is a table definition, it *must* declare itself the default
+(there is no way to access non-default tables anyhow).
+
+Each table definition also includes an *element type*, *initial length*, and
+optional *maximum length*.
+
+In the MVP, the only valid element type is `"function"`, but in the 
+[future](FutureFeatures.md#more-table-operators-and-types), more element
+types may be added.
+
+In the MVP, tables can only be resized via host-defined APIs (such as
+the JavaScript [`WebAssembly.Table.prototype.grow`](JS.md#webassemblytableprototypegrow)).
+A `grow_table` may be added in the [future](FutureFeatures.md#more-table-operators-and-types).
+In either case, table growth is guaranteed to fail if attempting to grow past
+the declared maximum. As with linear memory, when a maximum is declared,
+implementations *should* (non-normative) attempt to reserve virtual memory up to
+the maximum size. While failure to allocate the *initial* memory size is a
+runtime error, failure to reserve up to the *maximum* is not. When a maximum
+memory size is *not* declared, on architectures with limited virtual address
+space, engines should allocate only the initial size and reallocate on demand.
+
+## Elements section
+
+The intial contents of a tables' elements are sentinel values (that would throw
+if called). The *elements section* allows a module to initialize (at
+instantiation time) the elements of any imported or internally-defined table
+with any other definition in the module. This is symmetric to how the 
+[Data section](#data-section) allows a module to initialize the bytes
+of any imported or defined memory.
+
+Specifically, the elements section contains a possibly-empty array of
+*element segments* which specify the initial contents of fixed
+`(offset, length)` ranges of a given table, specified by its
+[table index](#table-index-space). The `length` is an integer constant value
+(defining the length of the given segment). The `offset` is an
+[initializer expression](#initializer-expression).
+
+## Function and Code sections
+
+A single logical function definition is defined in two sections: 
+ * the *function* section declares the signatures of each internal function
+   definition in the module;
+ * the *code* section contains the [function body](BinaryEncoding.md#function-bodies)
+   of each function declared by the function section.
+
+This split aids in streaming compilation by putting the function bodies,
+which constitute most of the byte size of the module, near the end so that all
+metadata necessary for recursive module loading and parallel compilation is
+available before compilation begins.
+
+## Definition Index Space
```

Comment:
Hm, can we somehow avoid this cumulative index space? It is weird that the same entity has different indices, depending on where it is referenced from. It also creates brittle forward compatibility -- imagine, for example, that we need to make type definitions exportable at some point (a rather likely scenario), what then?

Instead, we could e.g. distinguish the index/name spaces at the reference site. That will require an extra byte, but OTOH, the actual value will be smaller, which might give at least some space back in the LEB encoding (and compression).